unit-test builder should show the proper origin source file in errors (vitest browser mode)

### Command

test

### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

_No response_

### Description

When running tests, using the unit-test builder (default configuration with vitest), the errors look like:

```
Expected: "Hello!"
Received: "Hello, vitest-errors"

 ❯ spec-app-app.js:21:54
```

The source mentions a file that does not exist in the project, making it impossible to quickly navigate to it (the line line and columns are accurate though).

### Minimal Reproduction

```
npx @angular/cli@21 new vitest-errors --defaults
cd vitest-errors
```

Switch to browser mode in `angular.json`:
```
"test": {
  "builder": "@angular/build:unit-test",
  "options": {
    "browsers": ["ChromiumHeadless"]
  }
}
```

Add dependency:
```
npm i --save-dev --save-exact @vitest/browser-playwright
```

Change an assertion in `app.spec.ts` to make the tests fail, and run `ng test`

The error mentions `spec-app-app.js:21:54`, which is not the source file, and makes it impossible to quickly navigate to it.

### Exception or Error

```text
FAIL   vitest-errors (chromium)  src/app/app.spec.ts > App > should render title
AssertionError: expected 'Hello, vitest-errors' to contain 'Hello!'

Failure screenshot:
  - src/app/__screenshots__/app.spec.ts/App-should-render-title-1.png

Expected: "Hello!"
Received: "Hello, vitest-errors"

 ❯ spec-app-app.js:21:54
```

### Your Environment

```text
Angular CLI       : 21.0.3
Angular           : 21.0.4
Node.js           : 22.18.0
Package Manager   : npm 10.9.0
Operating System  : darwin arm64

┌───────────────────────────┬───────────────────┬───────────────────┐
│ Package                   │ Installed Version │ Requested Version │
├───────────────────────────┼───────────────────┼───────────────────┤
│ @angular/build            │ 21.0.3            │ ^21.0.3           │
│ @angular/cli              │ 21.0.3            │ ^21.0.3           │
│ @angular/common           │ 21.0.4            │ ^21.0.0           │
│ @angular/compiler         │ 21.0.4            │ ^21.0.0           │
│ @angular/compiler-cli     │ 21.0.4            │ ^21.0.0           │
│ @angular/core             │ 21.0.4            │ ^21.0.0           │
│ @angular/forms            │ 21.0.4            │ ^21.0.0           │
│ @angular/platform-browser │ 21.0.4            │ ^21.0.0           │
│ @angular/router           │ 21.0.4            │ ^21.0.0           │
│ rxjs                      │ 7.8.2             │ ~7.8.0            │
│ typescript                │ 5.9.3             │ ~5.9.2            │
│ vitest                    │ 4.0.15            │ ^4.0.8            │
└───────────────────────────┴───────────────────┴───────────────────┘
```

### Anything else relevant?

Note that this does not happen with jsdom:

```
 ❯ src/app/app.spec.ts:21:55
     19|     await fixture.whenStable();
     20|     const compiled = fixture.nativeElement as HTMLElement;
     21|     expect(compiled.querySelector('h1')?.textContent).toContain('Hello!');
       |                                                       ^
     22|   });
     23| });
```

--- a/packages/angular/build/src/builders/unit-test/runners/vitest/plugins.ts
+++ b/packages/angular/build/src/builders/unit-test/runners/vitest/plugins.ts
@@ -112,6 +112,12 @@ export async function createVitestConfigPlugin(
         delete config.plugins;
       }
 
+      // Add browser source map support
+      if (browser || testConfig?.browser?.enabled) {
+        projectPlugins.unshift(createSourcemapSupportPlugin());
+        setupFiles.unshift('virtual:source-map-support');
+      }
+
       const projectResolver = createRequire(projectSourceRoot + '/').resolve;
 
       const projectDefaults: UserWorkspaceConfig = {
@@ -308,6 +314,36 @@ export function createVitestPlugins(pluginOptions: PluginOptions): VitestPlugins
       },
     },
   ];
+}
+
+function createSourcemapSupportPlugin(): VitestPlugins[0] {
+  return {
+    name: 'angular:source-map-support',
+    enforce: 'pre',
+    resolveId(source) {
+      if (source.includes('virtual:source-map-support')) {
+        return '\0source-map-support';
+      }
+    },
+    async load(id) {
+      if (id !== '\0source-map-support') {
+        return;
+      }
+
+      const packageResolve = createRequire(__filename).resolve;
+      const supportPath = packageResolve('source-map-support/browser-source-map-support.js');
+
+      const content = await readFile(supportPath, 'utf-8');
+
+      // The `source-map-support` library currently relies on `this` being defined in the global scope.
+      // However, when running in an ESM environment, `this` is undefined.
+      // To workaround this, we patch the library to use `globalThis` instead of `this`.
+      return (
+        content.replaceAll(/this\.(define|sourceMapSupport|base64js)/g, 'globalThis.$1') +
+        '\n;globalThis.sourceMapSupport.install();'
+      );
+    },
+  };
 }
 
 async function generateCoverageOption(

--- a/tests/e2e/tests/vitest/browser-sourcemaps.ts
+++ b/tests/e2e/tests/vitest/browser-sourcemaps.ts
@@ -1,0 +1,40 @@
+import assert from 'node:assert/strict';
+import { applyVitestBuilder } from '../../utils/vitest';
+import { ng, noSilentNg } from '../../utils/process';
+import { installPackage } from '../../utils/packages';
+import { writeFile } from '../../utils/fs';
+import { stripVTControlCharacters } from 'node:util';
+
+export default async function (): Promise<void> {
+  await applyVitestBuilder();
+  await installPackage('playwright@1');
+  await installPackage('@vitest/browser-playwright@4');
+  await ng('generate', 'component', 'my-comp');
+
+  // Add a failing test to verify source map support
+  await writeFile(
+    'src/app/failing.spec.ts',
+    `
+      describe('Failing Test', () => {
+        it('should fail', () => {
+          expect(true).toBe(false);
+        });
+      });
+    `,
+  );
+
+  try {
+    await noSilentNg('test', '--no-watch', '--browsers', 'chromiumHeadless');
+    throw new Error('Expected "ng test" to fail.');
+  } catch (error: any) {
+    const stdout = stripVTControlCharacters(error.stdout || error.message);
+    // We expect the failure from failing.spec.ts
+    assert.match(stdout, /1 failed/, 'Expected 1 test to fail.');
+    // Check that the stack trace points to the correct file
+    assert.match(
+      stdout,
+      /\bsrc[\/\\]app[\/\\]failing\.spec\.ts:4:\d+/,
+      'Expected stack trace to point to the source file.',
+    );
+  }
+}

